engine/runtime/src/main/java/org/enso/interpreter/runtime/SerializationPool.java

```diff
@@ -0,0 +1,187 @@
+package org.enso.interpreter.runtime;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import org.enso.pkg.QualifiedName;
+
+final class SerializationPool {
+  private final TruffleCompilerContext context;
+
+  /**
+   * A set of the modules that are currently being serialized.
+   *
+   * <p>This set is accessed concurrently. This is safe as it is backed by a [[ConcurrentHashMap]]
+   * and is wrapped with the scala [[mutable.Set]] interface.
+   */
+  private final Map<QualifiedName, Boolean> isSerializing = new ConcurrentHashMap<>();
+
+  /**
+   * A map of the modules awaiting serialization to their associated tasks
+   *
+   * <p>This map is accessed concurrently.
+   */
+  private final Map<QualifiedName, Future<?>> isWaitingForSerialization = new ConcurrentHashMap<>();
+
+  /** The thread pool that handles serialization. */
+  private final ExecutorService pool;
+
+  /** all associated threads */
+  private final Set<Thread> threads = Collections.synchronizedSet(new HashSet<>());
+
+  SerializationPool(TruffleCompilerContext context) {
+    this.context = context;
+    this.pool =
+        Executors.newSingleThreadExecutor(
```

Comment:
Because then the code gets simpler to reason about?

Because there is no point in trying to execute two serializations in parallel?

Because the original code was a complicated mess?

Reply:
Fine. Since we are serializing IR during compilation that _might_ extend it slightly. Big "if" since I haven't measured it.